lib/Sema/CSDiag.cpp

```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
   return false;
 }
 
+/// Given a parameter type that may contain generic type params and an actual
+/// argument type, decide whether the param and actual arg have the same shape
+/// and equal fixed type portions, and return by reference each archetype and
+/// the matching portion of the actual arg type where that archetype appears.
+static bool findGenericSubstitutions(Type paramType, Type actualArgType,
+                                     SmallVector<ArchetypeType *, 4> &archetypes,
+                                     SmallVector<Type, 4> &substitutions) {
+  if (auto genericParam = paramType->getAs<GenericTypeParamType>())
+    paramType = genericParam->getDecl()->getArchetype();
```

Comment:
What if the paramType is a DependentMemberType? (Ie, an associated type reference).

Also getArchetype() is going away -- try this instead, which addresses both issues:

if (paramType->isTypeParameter())
    paramType = ArchetypeBuilder::mapTypeIntoContext(dc, paramType, nullptr);